common/setups/rasr/nn_system.py

```diff
@@ -0,0 +1,554 @@
+__all__ = ["NnArgs", "NnSystem"]
+
+import copy
+import itertools
+import logging
+import sys
+from typing import Dict, List, Optional, Tuple, Union
+
+# -------------------- Sisyphus --------------------
+
+import sisyphus.toolkit as tk
+import sisyphus.global_settings as gs
+
+# -------------------- Recipes --------------------
+
+import i6_core.am as am
+import i6_core.corpus as corpus_recipes
+import i6_core.meta as meta
+import i6_core.rasr as rasr
+import i6_core.returnn as returnn
+import i6_core.util as util
+
+from .rasr_system import RasrSystem
+
+from .util import (
+    RasrInitArgs,
+    ReturnnRasrDataInput,
+    OggZipHdfDataInput,
+    NnArgs,
+    RasrSteps,
+)
+
+from i6_private.users.luescher.util import (
+    load_multipath_from_bundle,
+    load_task_dependent_data,
+)
+
+# -------------------- Init --------------------
+
+Path = tk.setup_path(__package__)
+
+# -------------------- System --------------------
+
+
+class NnSystem(RasrSystem):
+    """
+    - 5 corpora types: train, devtrain, cv, dev and test
+        devtrain is a small split from the train set which is evaluated like
+        the cv but not used for error calculating. Since we can have different
+        datasubsets per subepoch, we do not caculate the tran score/error on
+        a consistent datasubset
+    - two training data settings: defined in returnn config or not
+    - 3 different types of decoding: returnn, rasr, rasr-label-sync
+    - 2 different lm: count, neural
+    - cv is dev for returnn training
+    - dev for lm param tuning
+    - test corpora for final eval
+
+    settings needed:
+    - am
+    - lm
+    - lexicon
+    - ce training
+    - ce recognition
+    - ce rescoring
+    - smbr training
+    - smbr recognition
+    - smbr rescoring
+    """
+
+    def __init__(
+        self,
+        returnn_root: Optional[str] = None,
+        returnn_python_home: Optional[str] = None,
+        returnn_python_exe: Optional[str] = None,
+    ):
+        super().__init__()
+
+        self.returnn_root = returnn_root or gs.RETURNN_ROOT
+        self.returnn_python_home = returnn_python_home or gs.RETURNN_PYTHON_HOME
+        self.return_python_exe = returnn_python_exe or gs.RETURNN_PYTHON_EXE
```

Comment:
My thinking was that `ReturnnTrainingJob` and `ReturnnRasrTrainingJob` also both require setting `RETURNN_*` in `gs`. Then I would rather have the setup fail as soon as possible.
But I did not think about setting `returnn_` at a later stage, e.g. via `step_args`.
I think you are right to leave this open so that the user has flexibility.

Reply:
There is no harm in having it fail here if it is unset. But then rather than throwing an `AttributeError` we should in this case assert that either is set.